File: frontend/src/api/scout.ts
import client from "./client";

export interface ScoutDashboardStats {
  players_viewed: number;
  saved_prospects: number;
  reports_written: number;
}

export interface RecentPlayerItem {
  id: string;
  first_name: string;
  last_name: string;
  position: string;
  nationality: string | null;
  club_name: string | null;
  age: number | null;
  market_value: number | null;
  last_viewed: string;
}

export interface SavedProspectSummary {
  player_id: string;
  first_name: string;
  last_name: string;
  position: string;
  nationality: string | null;
  club_name: string | null;
  age: number | null;
  saved_at: string;
}

export interface ScoutDashboardResponse {
  stats: ScoutDashboardStats;
  recently_viewed: RecentPlayerItem[];
  saved_prospects: SavedProspectSummary[];
}

export interface ScoutPlayerItem {
  id: string;
  first_name: string;
  last_name: string;
  position: string;
  age: number | null;
  nationality: string | null;
  club_id: string | null;
  club_name: string | null;
  market_value: number | null;
  status: string;
  is_saved: boolean;
}

export interface ScoutPlayersResponse {
  items: ScoutPlayerItem[];
  total: number;
  page: number;
  pages: number;
}

export interface ScoutSavedProspectItem {
  id: string;
  player_id: string;
  first_name: string;
  last_name: string;
  position: string;
  age: number | null;
  nationality: string | null;
  club_name: string | null;
  market_value: number | null;
  saved_at: string;
}

export interface ScoutReportItem {
  id: string;
  player_id: string | null;
  player_name: string;
  position: string;
  rating: number;
  status: string;
  notes: string | null;
  created_at: string;
  updated_at: string | null;
}

export interface PlayersQueryParams {
  page?: number;
  limit?: number;
  search?: string;
  position?: string;
  club_id?: string;
}

export interface PlayerDropdownItem {
  id: string;
  first_name: string;
  last_name: string;
  position: string;
}

export interface CreateReportPayload {
  player_id?: string;
  player_name: string;
  position: string;
  rating: number;
  status: string;
  notes?: string;
}

export interface UpdateReportPayload extends CreateReportPayload {}

export const scoutApi = {
  getDashboard: (): Promise<ScoutDashboardResponse> =>
    client.get("/scout/dashboard").then((r) => r.data),

  getPlayersForDropdown: (): Promise<PlayerDropdownItem[]> =>
    client.get("/scout/players/dropdown").then((r) => r.data),

  getPlayers: (params: PlayersQueryParams = {}): Promise<ScoutPlayersResponse> =>
    client.get("/scout/players", { params }).then((r) => r.data),

  recordView: (playerId: string): Promise<void> =>
    client.post(`/scout/players/${playerId}/view`).then(() => undefined),

  getSavedProspects: (): Promise<ScoutSavedProspectItem[]> =>
    client.get("/scout/saved-prospects").then((r) => r.data),

  saveProspect: (playerId: string): Promise<void> =>
    client.post(`/scout/saved-prospects/${playerId}`).then(() => undefined),

  unsaveProspect: (playerId: string): Promise<void> =>
    client.delete(`/scout/saved-prospects/${playerId}`).then(() => undefined),

  getReports: (): Promise<ScoutReportItem[]> =>
    client.get("/scout/reports").then((r) => r.data),

  createReport: (data: CreateReportPayload): Promise<ScoutReportItem> =>
    client.post("/scout/reports", data).then((r) => r.data),

  updateReport: (id: string, data: UpdateReportPayload): Promise<ScoutReportItem> =>
    client.put(`/scout/reports/${id}`, data).then((r) => r.data),

  deleteReport: (id: string): Promise<void> =>
    client.delete(`/scout/reports/${id}`).then(() => undefined),
};

File: backend/app/schemas/scout.py
from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, Field, field_validator


class ScoutDashboardStats(BaseModel):
    players_viewed: int
    saved_prospects: int
    reports_written: int


class RecentPlayerItem(BaseModel):
    id: str
    first_name: str
    last_name: str
    position: Optional[str]
    nationality: Optional[str]
    club_name: Optional[str]
    club_logo_url: Optional[str] = None
    age: Optional[int]
    market_value: Optional[int]
    last_viewed: datetime


class SavedProspectSummary(BaseModel):
    player_id: str
    first_name: str
    last_name: str
    position: Optional[str]
    nationality: Optional[str]
    club_name: Optional[str]
    club_logo_url: Optional[str] = None
    age: Optional[int]
    saved_at: datetime


class ScoutDashboardResponse(BaseModel):
    stats: ScoutDashboardStats
    recently_viewed: List[RecentPlayerItem]
    saved_prospects: List[SavedProspectSummary]
    new_players_since_last_visit: int = 0


class ScoutPlayerItem(BaseModel):
    id: str
    first_name: str
    last_name: str
    position: Optional[str]
    age: Optional[int]
    nationality: Optional[str]
    club_id: Optional[str]
    club_name: Optional[str]
    club_logo_url: Optional[str] = None
    market_value: Optional[int]
    status: str
    is_saved: bool


class ScoutPlayersResponse(BaseModel):
    items: List[ScoutPlayerItem]
    total: int
    page: int
    pages: int


class ScoutSavedProspectItem(BaseModel):
    id: str
    player_id: str
    first_name: str
    last_name: str
    position: Optional[str]
    age: Optional[int]
    nationality: Optional[str]
    club_name: Optional[str]
    club_logo_url: Optional[str] = None
    market_value: Optional[int]
    saved_at: datetime


class ScoutReportItem(BaseModel):
    id: str
    player_id: Optional[str]
    player_name: str
    position: str
    rating: int
    status: str
    notes: Optional[str]
    created_at: datetime
    updated_at: Optional[datetime]


_SCOUT_REPORT_STATUSES = {"draft", "submitted"}


class CreateScoutReportRequest(BaseModel):
    player_id: Optional[str] = None
    player_name: str = Field(min_length=1, max_length=200)
    position: str = Field(min_length=1, max_length=20)
    rating: int = Field(ge=1, le=100)
    status: str = "draft"
    notes: Optional[str] = Field(None, max_length=2000)

    @field_validator("status")
    @classmethod
    def validate_status(cls, v: str) -> str:
        if v not in _SCOUT_REPORT_STATUSES:
            raise ValueError("status must be 'draft' or 'submitted'")
        return v


class UpdateScoutReportRequest(BaseModel):
    player_id: Optional[str] = None
    player_name: str = Field(min_length=1, max_length=200)
    position: str = Field(min_length=1, max_length=20)
    rating: int = Field(ge=1, le=100)
    status: str = "draft"
    notes: Optional[str] = Field(None, max_length=2000)

    @field_validator("status")
    @classmethod
    def validate_status(cls, v: str) -> str:
        if v not in _SCOUT_REPORT_STATUSES:
            raise ValueError("status must be 'draft' or 'submitted'")
        return v


class PlayerDropdownItem(BaseModel):
    id: str
    first_name: str
    last_name: str
    position: Optional[str]
    club_name: Optional[str]
